src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs

```diff
@@ -293,6 +293,46 @@ internal SymbolicRegexNode<TSet> CreateConcat(SymbolicRegexNode<TSet>[] nodes, b
             return sr;
         }
 
+        /// <summary>
+        /// Make a concatenation of given nodes, if any regex is nothing then return nothing, eliminate
+        /// intermediate epsilons, if tryCreateFixedLengthMarker and length is fixed, add a fixed length
+        /// marker at the end.
+        /// </summary>
+        internal SymbolicRegexNode<TSet> CreateConcat(DoublyLinkedList<SymbolicRegexNode<TSet>> nodes, bool tryCreateFixedLengthMarker)
```

Comment:
I was thinking if we could avoid special casing `CreateConcat` for the new linked list. I the possibly reasonable way would be to create a `List` out of the `DoublyLinkedList` and use the existing method, but of course that incurs extra memory allocation. Only linear though.

Not sure if the reduced code is worth the extra allocation though.

Reply:
I can look into it. Another option is to have `DoublyLinkedList`  be `IEnumerable` and have `CreateConcat` take an `IEnumerable` instead.

Reply:
I added a variant of `CreateConcat` called `CreateConcatRev` that takes an `IEnumerable` that is already reversed. Creating an intermediate array each time seems too costly taking into account that it is being used a lot during the conversion. I also generalized `CalculateFixedLength` by making the argument be `IEnumerable`, and eliminated the variant of it using `DoublyLinkedList`.
Updated `CreateConcat(nodes, mark)` to could call into `CreateConcatRev(EnumerateArrayInReverse(nodes), mark)`  in order to avoid code duplication.
This also meant that I could make `DoublyLinkedList<T>.Node` private, as well as `_first` and `_last` are now private fields. Potentially the `DoublyLinkedList` class could now be moved to be a private class inside `RegexNodeConverter`, but, unless I hear differently, I leave is internal for now.